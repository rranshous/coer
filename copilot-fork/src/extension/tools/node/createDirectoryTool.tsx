/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

import * as l10n from '@vscode/l10n';
import type * as vscode from 'vscode';
import { IPromptPathRepresentationService } from '../../../platform/prompts/common/promptPathRepresentationService';
import { LanguageModelTextPart, LanguageModelToolResult, MarkdownString } from '../../../vscodeTypes';
import { ToolName } from '../common/toolNames';
import { ICopilotTool, ToolRegistry } from '../common/toolsRegistry';
import { formatUriForFileWidget, resolveToolInputPath } from './toolUtils';
import { IFileSystemService } from '../../../platform/filesystem/common/fileSystemService';

export interface ICreateDirectoryParams {
	dirPath: string;
}

export class CreateDirectoryTool implements ICopilotTool<ICreateDirectoryParams> {
	public static toolName = ToolName.CreateDirectory;

	constructor(
		@IPromptPathRepresentationService private readonly promptPathRepresentationService: IPromptPathRepresentationService,
		@IFileSystemService private readonly fileSystemService: IFileSystemService,
	) { }

	async invoke(options: vscode.LanguageModelToolInvocationOptions<ICreateDirectoryParams>, token: vscode.CancellationToken) {
		const uri = this.promptPathRepresentationService.resolveFilePath(options.input.dirPath);
		if (!uri) {
			throw new Error(`Invalid directory path`);
		}

		await this.fileSystemService.createDirectory(uri);

		return new LanguageModelToolResult([
			new LanguageModelTextPart(
				`Created directory at ${this.promptPathRepresentationService.getFilePath(uri)}`,
			)
		]);
	}

	prepareInvocation(options: vscode.LanguageModelToolInvocationPrepareOptions<ICreateDirectoryParams>, token: vscode.CancellationToken): vscode.ProviderResult<vscode.PreparedToolInvocation> {
		const uri = resolveToolInputPath(options.input.dirPath, this.promptPathRepresentationService);
		return {
			invocationMessage: new MarkdownString(l10n.t`Creating ${formatUriForFileWidget(uri)}`),
			pastTenseMessage: new MarkdownString(l10n.t`Created ${formatUriForFileWidget(uri)}`)
		};
	}
}

ToolRegistry.registerTool(CreateDirectoryTool);