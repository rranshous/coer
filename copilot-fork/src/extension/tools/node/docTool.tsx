/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

import type * as vscode from 'vscode';
import { IPromptPathRepresentationService } from '../../../platform/prompts/common/promptPathRepresentationService';
import { IWorkspaceService } from '../../../platform/workspace/common/workspaceService';
import { isLocation } from '../../../util/common/types';
import { URI } from '../../../util/vs/base/common/uri';
import { LanguageModelTextPart, LanguageModelToolResult } from '../../../vscodeTypes';
import { IBuildPromptContext } from '../../prompt/common/intents';
import { ToolName } from '../common/toolNames';
import { ToolRegistry } from '../common/toolsRegistry';
import { checkCancellation } from './toolUtils';


interface IDocInfoTool {
	readonly filePaths: string[];
}

class DocInfoTool implements vscode.LanguageModelTool<IDocInfoTool> {

	static readonly toolName = ToolName.DocInfo;

	private static _docTypeNames = new Map<string, string>([
		['typescript', 'TSDoc comment'],
		['typescriptreact', 'TSDoc comment'],
		['javascript', 'JSDoc comment'],
		['javascriptreact', 'JSDoc comment'],
		['python', 'docstring'],
	]);

	constructor(
		@IWorkspaceService private readonly workspaceService: IWorkspaceService,
		@IPromptPathRepresentationService private readonly _promptPathRepresentationService: IPromptPathRepresentationService,
	) { }

	async invoke(options: vscode.LanguageModelToolInvocationOptions<IDocInfoTool>, token: vscode.CancellationToken) {

		const docNames = new Set<string>();

		for (const filePath of options.input.filePaths) {
			const uri = this._promptPathRepresentationService.resolveFilePath(filePath);
			if (!uri) {
				continue;
			}
			const doc = await this.workspaceService.openTextDocumentAndSnapshot(uri);
			const docName: string = DocInfoTool._docTypeNames.get(doc.languageId) || 'documentation comment';
			docNames.add(docName);
		}

		checkCancellation(token);
		return new LanguageModelToolResult([
			new LanguageModelTextPart(`Please generate ${Array.from(docNames).join(', ')} for the respective files. ONLY add documentation and do not change the code.`)
		]);
	}

	async prepareInvocation(options: vscode.LanguageModelToolInvocationPrepareOptions<IDocInfoTool>, token: vscode.CancellationToken): Promise<vscode.PreparedToolInvocation> {
		return {
			presentation: 'hidden',
		};
	}

	async provideInput(promptContext: IBuildPromptContext): Promise<IDocInfoTool | undefined> {
		const seen = new Set<string>();

		const filePaths: string[] = [];
		const ranges: ([a: number, b: number, c: number, d: number] | undefined)[] = [];

		function addPath(path: string, range: vscode.Range | undefined) {
			if (!seen.has(path)) {
				seen.add(path);
				filePaths.push(path);
				ranges.push(range && [range.start.line, range.start.character, range.end.line, range.end.character]);
			}
		}

		for (const ref of promptContext.chatVariables) {
			if (URI.isUri(ref.value)) {
				addPath(this._promptPathRepresentationService.getFilePath(ref.value), undefined);
			} else if (isLocation(ref.value)) {
				addPath(this._promptPathRepresentationService.getFilePath(ref.value.uri), ref.value.range);
			}
		}

		if (promptContext.workingSet) {
			for (const file of promptContext.workingSet) {
				addPath(this._promptPathRepresentationService.getFilePath(file.document.uri), file.range);
			}
		}

		if (!filePaths.length) {
			// no context variables or working set
		}

		return {
			filePaths,
		};
	}
}

ToolRegistry.registerTool(DocInfoTool);
