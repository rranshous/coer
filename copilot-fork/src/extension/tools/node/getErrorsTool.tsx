/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

import * as l10n from '@vscode/l10n';
import { BasePromptElementProps, PromptElement, PromptElementProps } from '@vscode/prompt-tsx';
import type * as vscode from 'vscode';
import { ILanguageDiagnosticsService } from '../../../platform/languages/common/languageDiagnosticsService';
import { IPromptPathRepresentationService } from '../../../platform/prompts/common/promptPathRepresentationService';
import { IWorkspaceService } from '../../../platform/workspace/common/workspaceService';
import { getLanguage } from '../../../util/common/languages';
import { isLocation } from '../../../util/common/types';
import { CancellationToken } from '../../../util/vs/base/common/cancellation';
import { Disposable } from '../../../util/vs/base/common/lifecycle';
import { URI } from '../../../util/vs/base/common/uri';
import { IInstantiationService } from '../../../util/vs/platform/instantiation/common/instantiation';
import { DiagnosticSeverity, ExtendedLanguageModelToolResult, LanguageModelPromptTsxPart, MarkdownString, Range } from '../../../vscodeTypes';
import { findDiagnosticForSelectionAndPrompt } from '../../context/node/resolvers/fixSelection';
import { IBuildPromptContext } from '../../prompt/common/intents';
import { renderPromptElementJSON } from '../../prompts/node/base/promptRenderer';
import { Tag } from '../../prompts/node/base/tag';
import { DiagnosticContext, Diagnostics } from '../../prompts/node/inline/diagnosticsContext';
import { ToolName } from '../common/toolNames';
import { ICopilotTool, ToolRegistry } from '../common/toolsRegistry';
import { checkCancellation, formatUriForFileWidget, resolveToolInputPath } from './toolUtils';

interface IGetErrorsParams {
	filePaths: string[];
	// sparse array of ranges, as numbers because it goes through JSON
	ranges?: ([a: number, b: number, c: number, d: number] | undefined)[];
}

class GetErrorsTool extends Disposable implements ICopilotTool<IGetErrorsParams> {
	public static readonly toolName = ToolName.GetErrors;

	constructor(
		@IInstantiationService private readonly instantiationService: IInstantiationService,
		@ILanguageDiagnosticsService private readonly languageDiagnosticsService: ILanguageDiagnosticsService,
		@IWorkspaceService private readonly workspaceService: IWorkspaceService,
		@IPromptPathRepresentationService private readonly promptPathRepresentationService: IPromptPathRepresentationService,
	) {
		super();
	}

	async invoke(options: vscode.LanguageModelToolInvocationOptions<IGetErrorsParams>, token: CancellationToken) {
		const diagnostics = await Promise.all(options.input.filePaths.map(async (filePath, i): Promise<{ context: DiagnosticContext; uri: URI; diagnostics: vscode.Diagnostic[] }> => {
			const uri = resolveToolInputPath(filePath, this.promptPathRepresentationService);
			const range = options.input.ranges?.[i];
			if (!uri) {
				throw new Error(`Invalid input path ${filePath}`);
			}

			let diagnostics = range
				? findDiagnosticForSelectionAndPrompt(this.languageDiagnosticsService, uri, new Range(...range), undefined)
				: this.languageDiagnosticsService.getDiagnostics(uri);

			diagnostics = diagnostics.filter(d => d.severity <= DiagnosticSeverity.Warning);

			const document = await this.workspaceService.openTextDocumentAndSnapshot(uri);
			checkCancellation(token);

			return {
				context: { document, language: getLanguage(document) },
				diagnostics,
				uri,
			};
		}));

		checkCancellation(token);

		const result = new ExtendedLanguageModelToolResult([
			new LanguageModelPromptTsxPart(
				await renderPromptElementJSON(this.instantiationService, DiagnosticToolOutput, { diagnosticsGroups: diagnostics, maxDiagnostics: 50 }, options.tokenizationOptions, token)
			)
		]);

		const numDiagnostics = diagnostics.reduce((acc, { diagnostics }) => acc + diagnostics.length, 0);
		result.toolResultMessage = numDiagnostics === 0 ?
			new MarkdownString(l10n.t`Checked ${this.formatURIs(diagnostics.map(d => d.uri))}, no problems found`) :
			numDiagnostics === 1 ?
				new MarkdownString(l10n.t`Checked ${this.formatURIs(diagnostics.map(d => d.uri))}, 1 problem found`) :
				new MarkdownString(l10n.t`Checked ${this.formatURIs(diagnostics.map(d => d.uri))}, ${numDiagnostics} problems found`);
		return result;
	}

	prepareInvocation(options: vscode.LanguageModelToolInvocationPrepareOptions<IGetErrorsParams>, token: vscode.CancellationToken): vscode.ProviderResult<vscode.PreparedToolInvocation> {
		if (!options.input.filePaths?.length) {
			throw new Error('No file paths provided');
		}

		const uris = options.input.filePaths.map(filePath => resolveToolInputPath(filePath, this.promptPathRepresentationService));
		if (uris.some(uri => uri === undefined)) {
			throw new Error('Invalid file path provided');
		}

		return {
			invocationMessage: new MarkdownString(l10n.t`Checking ${this.formatURIs(uris)}`),
		};
	}

	private formatURIs(uris: URI[]): string {
		return uris.map(formatUriForFileWidget).join(', ');
	}

	async provideInput(promptContext: IBuildPromptContext): Promise<IGetErrorsParams | undefined> {
		const seen = new Set<string>();

		const filePaths: string[] = [];
		const ranges: ([a: number, b: number, c: number, d: number] | undefined)[] = [];

		function addPath(path: string, range: vscode.Range | undefined) {
			if (!seen.has(path)) {
				seen.add(path);
				filePaths.push(path);
				ranges.push(range && [range.start.line, range.start.character, range.end.line, range.end.character]);
			}
		}

		for (const ref of promptContext.chatVariables) {
			if (URI.isUri(ref.value)) {
				addPath(this.promptPathRepresentationService.getFilePath(ref.value), undefined);
			} else if (isLocation(ref.value)) {
				addPath(this.promptPathRepresentationService.getFilePath(ref.value.uri), ref.value.range);
			}
		}

		if (promptContext.workingSet) {
			for (const file of promptContext.workingSet) {
				addPath(this.promptPathRepresentationService.getFilePath(file.document.uri), file.range);
			}
		}

		if (!filePaths.length) {
			for (const [uri, diags] of this.languageDiagnosticsService.getAllDiagnostics()) {
				const path = this.promptPathRepresentationService.getFilePath(uri);
				if (diags.length) {
					let range = diags[0].range;
					for (let i = 1; i < diags.length; i++) {
						range = range.union(diags[i].range);
					}
					addPath(path, range);
				}
			}
		}

		return {
			filePaths,
			ranges
		};
	}
}

ToolRegistry.registerTool(GetErrorsTool);

interface IDiagnosticToolOutputProps extends BasePromptElementProps {
	diagnosticsGroups: { context: DiagnosticContext; uri: URI; diagnostics: vscode.Diagnostic[] }[];
	maxDiagnostics?: number;
}

export class DiagnosticToolOutput extends PromptElement<IDiagnosticToolOutputProps> {
	constructor(
		props: PromptElementProps<IDiagnosticToolOutputProps>,
		@IPromptPathRepresentationService private readonly promptPathRepresentationService: IPromptPathRepresentationService,
	) {
		super(props);
	}

	render() {
		if (!this.props.diagnosticsGroups.length) {
			return <>No errors found.</>;
		}

		let diagnosticsGroups = this.props.diagnosticsGroups;
		let limitMsg;
		if (typeof this.props.maxDiagnostics === 'number') {
			let remaining = this.props.maxDiagnostics;
			diagnosticsGroups = this.props.diagnosticsGroups.map(group => {
				if (remaining <= 0) {
					return { ...group, diagnostics: [] };
				}
				const take = Math.min(group.diagnostics.length, remaining);
				remaining -= take;
				return { ...group, diagnostics: group.diagnostics.slice(0, take) };
			});
			const totalDiagnostics = this.props.diagnosticsGroups.reduce((acc, group) => acc + group.diagnostics.length, 0);
			limitMsg = totalDiagnostics > this.props.maxDiagnostics
				? <>Showing first {this.props.maxDiagnostics} results out of {totalDiagnostics}<br /></>
				: undefined;
		}

		return <>
			{limitMsg}
			{diagnosticsGroups.map(d =>
				<Tag name='errors' attrs={{ path: this.promptPathRepresentationService.getFilePath(d.uri) }}>
					{d.diagnostics.length
						? <Diagnostics
							documentContext={d.context}
							diagnostics={d.diagnostics}
							includeRelatedInfos={false} // avoid blowing up the prompt #12655
						/>
						: 'No errors found'}
				</Tag>
			)}
		</>;
	}
}
