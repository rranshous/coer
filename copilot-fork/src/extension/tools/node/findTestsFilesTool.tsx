/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

import * as l10n from '@vscode/l10n';
import { BasePromptElementProps, PromptElement, PromptElementProps, PromptSizing } from '@vscode/prompt-tsx';
import type * as vscode from 'vscode';
import { TextDocumentSnapshot } from '../../../platform/editing/common/textDocumentSnapshot';
import { IIgnoreService } from '../../../platform/ignore/common/ignoreService';
import { IPromptPathRepresentationService } from '../../../platform/prompts/common/promptPathRepresentationService';
import { IWorkspaceService } from '../../../platform/workspace/common/workspaceService';
import { isLocation } from '../../../util/common/types';
import { CancellationToken } from '../../../util/vs/base/common/cancellation';
import { Disposable } from '../../../util/vs/base/common/lifecycle';

import { URI } from '../../../util/vs/base/common/uri';
import { IInstantiationService } from '../../../util/vs/platform/instantiation/common/instantiation';
import { ExtendedLanguageModelToolResult, LanguageModelPromptTsxPart, MarkdownString } from '../../../vscodeTypes';
import { IBuildPromptContext } from '../../prompt/common/intents';
import { isTestFile, TestFileFinder } from '../../prompt/node/testFiles';
import { renderPromptElementJSON } from '../../prompts/node/base/promptRenderer';
import { Tag } from '../../prompts/node/base/tag';

import { IParserService } from '../../../platform/parser/node/parserService';
import { TestDeps } from '../../intents/node/testIntent/testDeps';
import { ProjectedDocument } from '../../prompts/node/inline/summarizedDocument/summarizeDocument';
import { summarizeDocuments, SummarizeDocumentsItem } from '../../prompts/node/inline/summarizedDocument/summarizeDocumentHelpers';
import { CodeBlock } from '../../prompts/node/panel/safeElements';
import { ToolName } from '../common/toolNames';
import { ICopilotTool, ToolRegistry } from '../common/toolsRegistry';
import { checkCancellation, formatUriForFileWidget, resolveToolInputPath } from './toolUtils';


interface IFindTestFilesToolsParams {
	readonly filePaths: string[];
	// sparse array of ranges, as numbers because it goes through JSON
	readonly ranges?: ([a: number, b: number, c: number, d: number] | undefined)[];
}

class FindTestFilesTool extends Disposable implements ICopilotTool<IFindTestFilesToolsParams> {
	public static readonly toolName = ToolName.FindTestFiles;
	private readonly _testFileFinder: TestFileFinder;

	constructor(
		@IInstantiationService private readonly instantiationService: IInstantiationService,
		@IWorkspaceService private readonly workspaceService: IWorkspaceService,
		@IPromptPathRepresentationService private readonly promptPathRepresentationService: IPromptPathRepresentationService,
		@IIgnoreService private readonly ignoreService: IIgnoreService,
	) {
		super();
		this._testFileFinder = this.instantiationService.createInstance(TestFileFinder);
	}

	async invoke(options: vscode.LanguageModelToolInvocationOptions<IFindTestFilesToolsParams>, token: CancellationToken): Promise<vscode.LanguageModelToolResult> {

		let languageId: string | undefined;
		const inputUris: URI[] = [];
		const testFiles: IRelatedExistingFile[] = [];
		const anyTestFiles: IRelatedExistingFile[] = [];
		const srcFiles: IRelatedExistingFile[] = [];

		let lookForAnyTestFile: TextDocumentSnapshot | undefined;

		await Promise.all(options.input.filePaths.map(async filePath => {
			const uri = this.promptPathRepresentationService.resolveFilePath(filePath);
			if (!uri) {
				throw new Error(`Invalid input path ${filePath}`);
			}
			if (await this.ignoreService.isCopilotIgnored(uri)) {
				return;
			}
			const document = await this.workspaceService.openTextDocumentAndSnapshot(uri);
			inputUris.push(document.uri);
			if (languageId === undefined) {
				languageId = document.languageId;
			}

			if (token.isCancellationRequested) {
				return;
			}

			if (isTestFile(document)) {
				const srcFileUri = await this._testFileFinder.findFileForTestFile(document, token);
				if (srcFileUri && !await this.ignoreService.isCopilotIgnored(srcFileUri)) {
					const srcFileDocument = await this.workspaceService.openTextDocumentAndSnapshot(srcFileUri);
					srcFiles.push({ srcFile: srcFileUri, testFile: document.uri, document: srcFileDocument });
				}
			} else {
				const testFileUri = await this._testFileFinder.findTestFileForSourceFile(document, token);
				if (testFileUri) {
					const testFileDocument = await this.workspaceService.openTextDocumentAndSnapshot(testFileUri);
					testFiles.push({ srcFile: document.uri, testFile: testFileUri, document: testFileDocument });
				}
			}
		}));

		checkCancellation(token);

		if (testFiles.length === 0 && lookForAnyTestFile) {
			const testFileUri = await this._testFileFinder.findAnyTestFileForSourceFile(lookForAnyTestFile, token);
			if (testFileUri && !await this.ignoreService.isCopilotIgnored(testFileUri)) {
				const testFileDocument = await this.workspaceService.openTextDocumentAndSnapshot(testFileUri);
				anyTestFiles.push({ srcFile: lookForAnyTestFile.uri, testFile: testFileUri, document: testFileDocument });
			}
		}


		const nTestFilesFound = testFiles.length + srcFiles.length + anyTestFiles.length;

		const props: IFindTestFilesToolOutputProps = {
			languageId: languageId!,
			testFiles,
			srcFiles,
			anyTestFiles

		};
		const result = new ExtendedLanguageModelToolResult([
			new LanguageModelPromptTsxPart(
				await renderPromptElementJSON(this.instantiationService, FindTestFilesToolOutput, props, options.tokenizationOptions, token)
			)
		]);

		result.toolResultMessage = nTestFilesFound === 0 ?
			new MarkdownString(l10n.t`Checked ${this.formatURIs(inputUris)} for test related files, none found`) :
			nTestFilesFound === 1 ?
				new MarkdownString(l10n.t`Checked ${this.formatURIs(inputUris)}, 1 file found`) :
				new MarkdownString(l10n.t`Checked ${this.formatURIs(inputUris)}, ${nTestFilesFound} files found`);
		return result;
	}

	prepareInvocation(options: vscode.LanguageModelToolInvocationPrepareOptions<IFindTestFilesToolsParams>, token: vscode.CancellationToken): vscode.ProviderResult<vscode.PreparedToolInvocation> {
		if (!options.input.filePaths?.length) {
			throw new Error('Invalid input');
		}

		const uris = options.input.filePaths.map(filePath => resolveToolInputPath(filePath, this.promptPathRepresentationService));
		if (uris.some(uri => uri === undefined)) {
			throw new Error('Invalid input');
		}

		return {
			invocationMessage: new MarkdownString(l10n.t`Checking ${this.formatURIs(uris)}`),
		};
	}

	private formatURIs(uris: URI[]): string {
		return uris.map(formatUriForFileWidget).join(', ');
	}

	async provideInput(promptContext: IBuildPromptContext): Promise<IFindTestFilesToolsParams | undefined> {
		const seen = new Set<string>();

		const filePaths: string[] = [];
		const ranges: ([a: number, b: number, c: number, d: number] | undefined)[] = [];

		function addPath(path: string, range: vscode.Range | undefined) {
			if (!seen.has(path)) {
				seen.add(path);
				filePaths.push(path);
				ranges.push(range && [range.start.line, range.start.character, range.end.line, range.end.character]);
			}
		}

		for (const ref of promptContext.chatVariables) {
			if (URI.isUri(ref.value)) {
				addPath(this.promptPathRepresentationService.getFilePath(ref.value), undefined);
			} else if (isLocation(ref.value)) {
				addPath(this.promptPathRepresentationService.getFilePath(ref.value.uri), ref.value.range);
			}
		}

		if (promptContext.workingSet) {
			for (const file of promptContext.workingSet) {
				addPath(this.promptPathRepresentationService.getFilePath(file.document.uri), file.range);
			}
		}

		if (!filePaths.length) {
			// no context variables or working set
		}

		return {
			filePaths,
			ranges
		};
	}
}

interface IRelatedExistingFile {
	readonly srcFile: URI;
	readonly testFile: URI;
	readonly document: TextDocumentSnapshot;
}

ToolRegistry.registerTool(FindTestFilesTool);

interface IFindTestFilesToolOutputProps extends BasePromptElementProps {
	readonly languageId: string;
	readonly testFiles: IRelatedExistingFile[];
	readonly anyTestFiles: IRelatedExistingFile[];
	readonly srcFiles: IRelatedExistingFile[];
}

class FindTestFilesToolOutput extends PromptElement<IFindTestFilesToolOutputProps> {
	constructor(
		props: PromptElementProps<IFindTestFilesToolOutputProps>,
		@IParserService private readonly parserService: IParserService,
	) {
		super(props);
	}

	async render(_state: void, sizing: PromptSizing) {
		if (this.props.testFiles.length === 0 && this.props.srcFiles.length === 0 && this.props.anyTestFiles.length === 0) {
			return <>No test related files found.</>;
		}

		const tokenSizing = Math.min(sizing.tokenBudget, 32_000);

		const documentData: SummarizeDocumentsItem[] = [...this.props.testFiles, ...this.props.anyTestFiles, ...this.props.srcFiles].map(info => ({ document: info.document, formattingOptions: undefined, selection: undefined }));
		const docs = await summarizeDocuments(this.parserService, documentData, tokenSizing);

		let index = 0;
		return <>
			<TestDeps languageId={this.props.languageId} />

			The following files are useful when writing tests: <br />
			{
				this.props.testFiles.map(info => <RelatedTestDescription info={info} projectedDoc={docs[index++]} />)
			}
			{
				this.props.anyTestFiles.map(info => <ExampleTestDescription info={info} projectedDoc={docs[index++]} />)
			}
			{
				this.props.srcFiles.map(info => <RelatedSourceDescription info={info} projectedDoc={docs[index++]} />)
			}
		</ >;
	}
}

interface IRelatedSourceDescriptionProps extends BasePromptElementProps {
	readonly projectedDoc: ProjectedDocument;
	readonly info: IRelatedExistingFile;
}


class RelatedSourceDescription extends PromptElement<IRelatedSourceDescriptionProps> {
	constructor(
		props: PromptElementProps<IRelatedSourceDescriptionProps>,
		@IPromptPathRepresentationService private readonly promptPathRepresentationService: IPromptPathRepresentationService,
	) {
		super(props);
	}


	render(state: void, sizing: PromptSizing) {
		const document = this.props.info.document;
		return <Tag name='relatedSource'>
			The test file {this.promptPathRepresentationService.getFilePath(this.props.info.testFile)} contains tests for the following file:<br />
			<CodeBlock code={this.props.projectedDoc.text} uri={document.uri} languageId={document.languageId} includeFilepath={true} /><br />
		</Tag>;
	}
}


interface IRelatedTestDescriptionProps extends BasePromptElementProps {
	readonly projectedDoc: ProjectedDocument;
	readonly info: IRelatedExistingFile;
}


class RelatedTestDescription extends PromptElement<IRelatedTestDescriptionProps> {
	constructor(
		props: PromptElementProps<IRelatedTestDescriptionProps>,
		@IPromptPathRepresentationService private readonly promptPathRepresentationService: IPromptPathRepresentationService,
	) {
		super(props);
	}

	render(state: void, sizing: PromptSizing) {
		const document = this.props.info.document;
		return <Tag name='relatedTest'>
			Tests for {this.promptPathRepresentationService.getFilePath(this.props.info.srcFile)} can go into the following existing file:<br />
			<CodeBlock code={this.props.projectedDoc.text} uri={document.uri} languageId={document.languageId} includeFilepath={true} /><br />
		</Tag>;
	}
}

interface IExampleTestDescriptionProps extends BasePromptElementProps {
	readonly info: IRelatedExistingFile;
	readonly projectedDoc: ProjectedDocument;
}


class ExampleTestDescription extends PromptElement<IExampleTestDescriptionProps> {
	constructor(
		props: PromptElementProps<IExampleTestDescriptionProps>,
	) {
		super(props);
	}

	render(state: void, sizing: PromptSizing) {
		return <Tag name='sampleTest'>
			This is a sample test file:<br />
			<CodeBlock code={this.props.projectedDoc.text} uri={this.props.info.document.uri} languageId={this.props.info.document.languageId} includeFilepath={true} /><br />
		</Tag>;
	}
}

