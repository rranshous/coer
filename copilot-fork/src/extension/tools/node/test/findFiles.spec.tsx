/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

import { afterEach, beforeEach, expect, suite, test } from 'vitest';
import type * as vscode from 'vscode';
import { RelativePattern } from '../../../../platform/filesystem/common/fileTypes';
import { AbstractSearchService, ISearchService } from '../../../../platform/search/common/searchService';
import { ITestingServicesAccessor, TestingServiceCollection } from '../../../../platform/test/node/services';
import { TestWorkspaceService } from '../../../../platform/test/node/testWorkspaceService';
import { IWorkspaceService } from '../../../../platform/workspace/common/workspaceService';
import { CancellationToken } from '../../../../util/vs/base/common/cancellation';
import { isWindows } from '../../../../util/vs/base/common/platform';
import { URI } from '../../../../util/vs/base/common/uri';
import { SyncDescriptor } from '../../../../util/vs/platform/instantiation/common/descriptors';
import { IInstantiationService } from '../../../../util/vs/platform/instantiation/common/instantiation';
import { createExtensionUnitTestingServices } from '../../../test/node/services';
import { CopilotToolMode } from '../../common/toolsRegistry';
import { FindFilesTool, IFindFilesToolParams } from '../findFilesTool';

suite('FindFiles', () => {
	let accessor: ITestingServicesAccessor;
	let collection: TestingServiceCollection;

	const workspaceFolder = isWindows ? 'c:\\test\\workspace' : '/test/workspace';

	beforeEach(() => {
		collection = createExtensionUnitTestingServices();
		collection.define(IWorkspaceService, new SyncDescriptor(TestWorkspaceService, [[URI.file(workspaceFolder)]]));
	});

	afterEach(() => {
		accessor.dispose();
	});

	function setup(expected: vscode.GlobPattern) {
		const patterns: vscode.GlobPattern[] = [expected];
		if (typeof expected === 'string' && !expected.endsWith('/**')) {
			patterns.push(expected + '/**');
		} else if (typeof expected !== 'string' && !expected.pattern.endsWith('/**')) {
			patterns.push(new RelativePattern(expected.baseUri, expected.pattern + '/**'));
		}
		collection.define(ISearchService, new TestSearchService(patterns));
		accessor = collection.createTestingAccessor();
	}

	test('passes through simple query', async () => {
		setup('test/**/*.ts');

		const tool = accessor.get(IInstantiationService).createInstance(FindFilesTool);
		await tool.invoke({ input: { query: 'test/**/*.ts' }, toolInvocationToken: null!, }, CancellationToken.None);
	});

	test('handles absolute path with glob', async () => {
		setup(new RelativePattern(URI.file(workspaceFolder), 'test/**/*.ts'));

		const tool = accessor.get(IInstantiationService).createInstance(FindFilesTool);
		await tool.invoke({ input: { query: `${workspaceFolder}/test/**/*.ts` }, toolInvocationToken: null!, }, CancellationToken.None);
	});

	test('handles absolute path to folder', async () => {
		setup(new RelativePattern(URI.file(workspaceFolder), ''));

		const tool = accessor.get(IInstantiationService).createInstance(FindFilesTool);
		await tool.invoke({ input: { query: workspaceFolder }, toolInvocationToken: null!, }, CancellationToken.None);
	});

	suite('resolveInput', () => {
		beforeEach(() => {
			setup('hello');
		});

		async function testIt(input: IFindFilesToolParams, context: CopilotToolMode) {
			const tool = accessor.get(IInstantiationService).createInstance(FindFilesTool);
			const resolved = await tool.resolveInput(input, null!, context);
			expect(resolved).toMatchSnapshot();
		}

		test('resolveInput with FullContext and no maxResults', async () => {
			await testIt({ query: 'hello' }, CopilotToolMode.FullContext);
		});

		test('resolveInput with FullContext and maxResults < 200', async () => {
			await testIt({ query: 'hello', maxResults: 50 }, CopilotToolMode.FullContext);
		});

		test('resolveInput with FullContext and maxResults > 200', async () => {
			await testIt({ query: 'hello', maxResults: 300 }, CopilotToolMode.FullContext);
		});

		test('resolveInput with PartialContext and no maxResults', async () => {
			await testIt({ query: 'hello' }, CopilotToolMode.PartialContext);
		});

		test('resolveInput with PartialContext and maxResults defined', async () => {
			await testIt({ query: 'hello', maxResults: 123 }, CopilotToolMode.PartialContext);
		});
	});
});

class TestSearchService extends AbstractSearchService {
	constructor(private readonly expectedPattern: vscode.GlobPattern | vscode.GlobPattern[]) {
		super();
	}

	override async findTextInFiles(query: vscode.TextSearchQuery, options: vscode.FindTextInFilesOptions, progress: vscode.Progress<vscode.TextSearchResult>, token: vscode.CancellationToken): Promise<vscode.TextSearchComplete> {
		throw new Error('Method not implemented.');
	}

	override findTextInFiles2(query: vscode.TextSearchQuery2, options?: vscode.FindTextInFilesOptions2, token?: vscode.CancellationToken): vscode.FindTextInFilesResponse {
		throw new Error('Method not implemented.');
	}

	override async findFiles(filePattern: vscode.GlobPattern | vscode.GlobPattern[], options?: vscode.FindFiles2Options | undefined, token?: vscode.CancellationToken | undefined): Promise<vscode.Uri[]> {
		expect(filePattern).toEqual(this.expectedPattern);
		return [];
	}
}