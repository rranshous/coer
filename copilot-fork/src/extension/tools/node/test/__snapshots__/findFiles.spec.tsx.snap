// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FindFiles > resolveInput > resolveInput with FullContext and maxResults < 200 1`] = `
{
  "maxResults": 200,
  "query": "**/hello",
}
`;

exports[`FindFiles > resolveInput > resolveInput with FullContext and maxResults > 200 1`] = `
{
  "maxResults": 300,
  "query": "**/hello",
}
`;

exports[`FindFiles > resolveInput > resolveInput with FullContext and no maxResults 1`] = `
{
  "maxResults": 200,
  "query": "**/hello",
}
`;

exports[`FindFiles > resolveInput > resolveInput with PartialContext and maxResults defined 1`] = `
{
  "maxResults": 123,
  "query": "**/hello",
}
`;

exports[`FindFiles > resolveInput > resolveInput with PartialContext and no maxResults 1`] = `
{
  "maxResults": 20,
  "query": "**/hello",
}
`;
